pkg/core/cloudstorage/presign_upload_request.go

```diff
@@ -0,0 +1,81 @@
+package cloudstorage
+
+import (
+	"fmt"
+	"mime"
+	"net/http"
+	"strconv"
+
+	"github.com/skygeario/skygear-server/pkg/core/uuid"
+)
+
+type PresignUploadRequest struct {
+	Prefix  string                 `json:"prefix,omitempty"`
+	Access  AccessType             `json:"access,omitempty"`
+	Headers map[string]interface{} `json:"headers"`
+}
+
+func (r *PresignUploadRequest) ContentLength() (contentLength int) {
+	if s, ok := r.Headers["content-length"].(string); ok {
+		i, err := strconv.Atoi(s)
+		if err == nil {
+			contentLength = i
+		}
+	}
+	return
+}
+
+func (r *PresignUploadRequest) SetDefaultValue() {
+	if r.Access == "" {
+		r.Access = AccessTypeDefault
+	}
+	if _, ok := r.Headers["content-type"]; !ok {
+		r.Headers["content-type"] = "application/octet-stream"
+	}
+}
+
+func (r *PresignUploadRequest) DeriveAssetName() (assetName string, err error) {
+	// Derive file extension
+	contentType, ok := r.Headers["content-type"].(string)
+	var ext string
+	if ok {
+		exts, err := mime.ExtensionsByType(contentType)
```

Comment:
Seems `mime` only has a small set of built-in types, we need to install `mailcap` on alpine to get better extension list.